File: inno-erp/inno_erp/inno_stock/overrides/stock_entry/stock_entry.py
import frappe
from erpnext.stock.doctype.stock_entry.stock_entry import make_stock_in_entry as erpnext_make_stock_in_entry
from frappe.model.mapper import get_mapped_doc
from frappe.utils import flt


@frappe.whitelist()
def make_stock_in_entry(source_name, target_doc=None):
	entry = frappe.db.get_value(
		"Stock Entry", source_name, ["stock_entry_type", "add_to_transit"], as_dict=True
	)
	if entry.stock_entry_type != "Material Transfer" or entry.add_to_transit is False:
		return erpnext_make_stock_in_entry(source_name, target_doc)

	def set_missing_values(source, target):
		target.stock_entry_type = "Material Transfer"
		target.set_missing_values()

		if not frappe.db.get_single_value("Stock Settings", "use_serial_batch_fields"):
			target.make_serial_and_batch_bundle_for_transfer()

	def update_item(source_doc, target_doc, source_parent):
		if source_doc.material_request_item and source_doc.material_request:
			add_to_transit = frappe.db.get_value("Stock Entry", source_name, "add_to_transit")
			if add_to_transit:
				warehouse = frappe.get_value(
					"Material Request Item", source_doc.material_request_item, "warehouse"
				)
				target_doc.t_warehouse = warehouse

		target_doc.s_warehouse = source_parent.to_warehouse
		target_doc.t_warehouse = source_parent.custom_end_warehouse
		target_doc.qty = source_doc.qty - source_doc.transferred_qty

	doclist = get_mapped_doc(
		"Stock Entry",
		source_name,
		{
			"Stock Entry": {
				"doctype": "Stock Entry",
				"field_map": {
					"name": "outgoing_stock_entry",
					"to_warehouse": "from_warehouse",
					"custom_end_warehouse": "to_warehouse",
				},
				"validation": {"docstatus": ["=", 1]},
			},
			"Stock Entry Detail": {
				"doctype": "Stock Entry Detail",
				"field_map": {
					"name": "ste_detail",
					"parent": "against_stock_entry",
					"serial_no": "serial_no",
					"batch_no": "batch_no",
				},
				"postprocess": update_item,
				"condition": lambda doc: flt(doc.qty) - flt(doc.transferred_qty) > 0.00001,
			},
		},
		target_doc,
		set_missing_values,
	)

	return doclist

File: inno-erp/inno_erp/inno_stock/overrides/stock_entry/stock_entry.js
frappe.ui.form.on("Stock Entry", {
	onload: (frm) => {
		frm.set_df_property("custom_end_warehouse", "hidden", 1);
	},
	refresh: (frm) => {
		frm.trigger("toggle_display_warehouses");
		frm.refresh_fields("from_warehouse");
		frm.refresh_fields("to_warehouse");
	},
	stock_entry_type: (frm) => {
		const doc = frm.doc;
		const section_container = frm.fields_dict.to_warehouse.section.df.fieldname;
		frm.set_value("add_to_transit", 0);

		frm.trigger("toggle_display_warehouses");
		frm.trigger("reset_reference_warehouses");

		if (
			frm.is_new() &&
			doc.outgoing_stock_entry == null &&
			doc.stock_entry_type === "Material Transfer"
		) {
			frm.set_value("add_to_transit", 1);
			if (frm.fields_dict[section_container].collapse_link.hasClass("collapsed")) {
				frm.fields_dict[section_container].collapse();
			}
		}
	},

	before_save: (frm) => {
		for (const item of frm.doc.items) {
			frappe.utils.set_branch_cost_center_for_item(frm, "Stock Entry Detail", item.name);
		}
	},
	add_to_transit: (frm) => {
		frm.trigger("toggle_display_warehouses");
	},
	toggle_display_warehouses: (frm) => {
		const doc = frm.doc;

		if (!doc.add_to_transit) {
			frm.set_df_property("custom_end_warehouse", "hidden", 1);
			return;
		}

		if (doc.outgoing_stock_entry != null && doc.stock_entry_type === "Material Transfer") {
			frm.set_df_property("custom_end_warehouse", "hidden", 1);
		}

		if (doc.outgoing_stock_entry == null && doc.stock_entry_type === "Material Transfer") {
			frm.set_df_property("custom_end_warehouse", "hidden", 0);
		}
	},

	reset_reference_warehouses: (frm) => {
		const doc = frm.doc;

		frm.set_value("custom_end_warehouse", null);
		frm.set_value("to_warehouse", null);
		frm.set_value("from_warehouse", null);

		if (doc.stock_entry_type !== "Material Transfer") {
			frm.set_query("to_warehouse", () => {
				return {};
			});
		}

		if (doc.items && doc.items.length > 0) {
			for (let item in doc.items) {
				frappe.model.set_value(
					doc.items[item].doctype,
					doc.items[item].name,
					"s_warehouse",
					null
				);
				frappe.model.set_value(
					doc.items[item].doctype,
					doc.items[item].name,
					"t_warehouse",
					null
				);
			}
		}
	},
});
